pelican_url/discovery_test.go

```diff
@@ -0,0 +1,294 @@
+/***************************************************************
+*
+* Copyright (C) 2024, Pelican Project, Morgridge Institute for Research
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you
+* may not use this file except in compliance with the License.  You may
+* obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+***************************************************************/
+
+package pelican_url
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/sirupsen/logrus/hooks/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// Spin up a discovery server for testing purposes
+func getTestDiscoveryServer(t *testing.T) *httptest.Server {
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == "/.well-known/pelican-configuration" {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			_, err := w.Write([]byte(`{
+				"director_endpoint": "https://director.com",
+				"namespace_registration_endpoint": "https://registration.com",
+				"broker_endpoint": "https://broker.com",
+				"jwks_uri": "https://tokens.com"
+			}`))
+			assert.NoError(t, err)
+		} else {
+			http.NotFound(w, r)
+		}
+	}
+	server := httptest.NewTLSServer(http.HandlerFunc(handler))
+	return server
+}
+
+func TestSetOSDFDiscoveryHost(t *testing.T) {
+	tests := []struct {
+		host     string
+		expected string
+	}{
+		{"director.org", "director.org"},
+		{"director.org:1234", "director.org:1234"},
+		{"https://director.org", "director.org"},
+		{"https://director.org:1234", "director.org:1234"},
+	}
+
+	// Note we can't run these in parallel by wrapping in test.Run because they'll stomp on each other due to the global variable
+	for _, test := range tests {
+		_, err := SetOsdfDiscoveryHost(test.host)
+		require.NoError(t, err)
+		assert.Equal(t, test.expected, OsdfDiscoveryHost)
+	}
+}
+
+func TestDiscoverFederation(t *testing.T) {
+	discServer := getTestDiscoveryServer(t)
+	defer discServer.Close()
+	discUrl, err := url.Parse(discServer.URL)
+	require.NoError(t, err)
+
+	t.Run("TestVanilla", func(t *testing.T) {
```

Comment:
The original TestDiscoverFederation also has an invalid URL test. Is it removed because the PelicanUrl validation should ensure that before Discover Federation is run?

Reply:
In theory, yeah. The problem is that `DiscoverFederation` is still exported, so I should add a few error checks. Thanks for catching that.